pygsti/tools/matrixtools.py

```diff
@@ -73,7 +105,15 @@ def is_pos_def(mx, tol=1e-9):
     bool
         True if mx is positive-semidefinite, otherwise False.
     """
-    evals = _np.linalg.eigvals(mx)
+    if not is_hermitian(mx, tol):
+        return False
+    if attempt_cholesky:
+        try:
+            _ = _spl.cholesky(mx)
+            return True # Cholesky succeeded
+        except _spl.LinAlgError:
+            pass # we fall back on eigenvalue decomposition
+    evals = _np.linalg.eigvalsh(mx)
```

Comment:
The property of being positive semidefinite is generally ascribed to Hermitian matrices, since the value of the quadratic form $x^\dagger A x = x^\dagger((A + A^{\dagger})/2)x$ only depends on the Hermitian part of $A$. The original implementation didn't check that, and so it sometimes reported that asymmetric matrices with positive eigenvalues were positive definite.

The new implementation starts with a check that the matrix is Hermitian. If we do have a Hermitian matrix then the fastest way to check if it's positive definite is a Cholesky decomposition. However, Cholesky will fail on matrices that are only positive semidefinite, so there is no point in trying Cholesky if we expect ``is_pos_def`` will regularly handle singular matrices. I've set the default behavior for ``attempt_cholesky`` to False.

Reply:
A couple of things:

1. Should we ensure the docstring is consistent? It says PD in the first line, but return says PSD.
2. Both implementations are clearly for PSD. Should the function name be `is_pos_semidef` instead? Pro is that it is more accurate, con is that it is an API change. While this is in matrixtools and therefore could have moderate/high user exposure, it seems like the only use is in `is_valid_density_mx` below (which also seems like it has low usage), so maybe the API change would go mostly unnoticed?

Reply:
In the context of quantum information I find people are lax about positive definite versus positive semidefinite, so I think the function name is fine. But I'll definitely change the docstring to be consistent about really checking that a matrix is PSD.